plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/state/MapboxViewportStateBase.kt

```diff
@@ -0,0 +1,156 @@
+package com.mapbox.maps.plugin.viewport.experimental.state
+
+import android.animation.Animator
+import android.animation.AnimatorSet
+import android.animation.ValueAnimator
+import androidx.core.animation.doOnEnd
+import com.mapbox.maps.plugin.animation.CameraAnimationsPlugin
+import com.mapbox.maps.plugin.viewport.DEFAULT_FRAME_TRANSITION_MAX_DURATION_MS
+import com.mapbox.maps.plugin.viewport.experimental.ViewportPlugin
+import com.mapbox.maps.plugin.viewport.experimental.data.ViewportData
+import com.mapbox.maps.plugin.viewport.experimental.data.ViewportDataSource
+import com.mapbox.maps.plugin.viewport.experimental.data.ViewportDataSourceUpdateObserver
+import com.mapbox.maps.plugin.viewport.experimental.transition.TransitionEndListener
+import com.mapbox.maps.plugin.viewport.experimental.transition.ViewportStateTransition
+import com.mapbox.maps.plugin.viewport.experimental.transition.ViewportTransitionOptions
+import java.util.concurrent.CopyOnWriteArrayList
+
+abstract class MapboxViewportStateBase : ViewportState {
+  protected var frameTransitionOptions =
+    ViewportTransitionOptions.Builder().maxDuration(DEFAULT_FRAME_TRANSITION_MAX_DURATION_MS)
+      .build()
+  protected var runningAnimation: AnimatorSet? = null
+  protected val transitionEndListeners = CopyOnWriteArrayList<TransitionEndListener>()
+
+  abstract val stateTransition: ViewportStateTransition
+  abstract val cameraPlugin: CameraAnimationsPlugin
+
+  abstract val dataSource: ViewportDataSource
+
+  private val dataSourceUpdateObserver =
+    ViewportDataSourceUpdateObserver { viewportData ->
+      if (!isTransitioning) {
+        updateFrame(viewportData, true)
+      }
+    }
+
+  /**
+   * Boolean flag that indicate if it is in a transition state from previous state.
+   */
+  override var isTransitioning: Boolean = false
+    protected set
+
+  /**
+   * Start transition to this state.
+   *
+   * @param stateTransitionOptions defines the transition options for the state transition from previous state.
+   * @param frameTransitionOptions defines the transition options for the continues frame update after the state transition.
+   * @param transitionEndListener listener that will be called after the state transition has finished.
+   */
+  override fun transitionToState(
+    stateTransitionOptions: ViewportTransitionOptions,
+    frameTransitionOptions: ViewportTransitionOptions,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    dataSource.registerUpdateObserver(dataSourceUpdateObserver)
+    startAnimation(
+      stateTransition.transitionToState(dataSource.getViewportData().camera, stateTransitionOptions)
+        .apply {
+          addListener(
+            createTransitionListener(frameTransitionOptions)
+          )
+        },
+      instant = false,
+      transitionEndListener
+    )
```

Comment:
This assumes that the dataSource will have a valid camera synchronously after we register an update observer. Isn't it possible that the dataSource might be inactive prior to when we need to transition into it and that the camera will not be valid right away?

Reply:
Good point, we need to validate this, what do you expect it to behave when there's no valid camera updates yet from the data source, shall we then wait until there's one update?

Reply:
This is a question that we had a problem answering with the `NavigationCamera` and settled on the data source having a default value of null island. It might be good to experiment with having a nullable `ViewportData` and running the transition only after a first update is generated since it could result in a better/clearer dev experience.

Reply:
In the latest iteration, we have made the viewport data observable, Transitions can observe the incoming data points and only start animation when the first data point arrives. The observer will also observe newer updates, giving the Transition needed information to adjust the animation's target position over the air.